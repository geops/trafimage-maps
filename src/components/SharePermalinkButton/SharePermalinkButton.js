import React, { useState, useEffect, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { MdClose } from 'react-icons/md';
import { FaLink } from 'react-icons/fa';
import PropTypes from 'prop-types';
import PermalinkInput from '@geops/react-ui/components/PermalinkInput';
import Button from '@geops/react-ui/components/Button';

const propTypes = {
  title: PropTypes.string,
  icon: PropTypes.node,
  className: PropTypes.string,
};

const defaultProps = {
  title: 'Permalink generieren',
  icon: <FaLink focusable={false} />,
  className: 'wkp-share-permalink-button',
};

function SharePermalinkButton({ icon, title, className }) {
  const [showTooltip, setShowTooltip] = useState(null);
  const [positionTooltip, setPositionTooltip] = useState();
  const { t } = useTranslation();
  const ref = useRef();

  // Close the tooltip when clicking outside the tooltip
  useEffect(() => {
    const onDocClick = e => {
      // If the click comes from an element of Autocomplete, don't close the list.
      if (ref && ref.current && ref.current.contains(e.target)) {
        return;
      }
      setShowTooltip(false);
    };
    if (showTooltip) {
      document.addEventListener('click', onDocClick);
    }
    return function cleanup() {
      document.removeEventListener('click', onDocClick);
    };
  }, [showTooltip]);

  return (
    <div className={className} ref={ref}>
      <Button
        title={t(title)}
        onClick={e => {
          setPositionTooltip(e.currentTarget.getBoundingClientRect());
          setShowTooltip(!showTooltip);
        }}
      >
        {icon}
      </Button>
      {showTooltip && positionTooltip && (
        <div
          className="wkp-tooltip"
          style={{
            left: positionTooltip.left + 30,
            top: positionTooltip.top,
          }}
        >
          <div className="wkp-arrow-left" />
          <Button
            className="wkp-close-bt"
            onClick={() => {
              setShowTooltip(false);
            }}
          >
            <MdClose focusable={false} />
          </Button>
          <PermalinkInput value={window.location.href} />
          <p>
            {t(
              'Sie k√∂nnen auch den Link aus der Adresszeile des Browsers kopieren.',
            )}
          </p>
        </div>
      )}
    </div>
  );
}

SharePermalinkButton.propTypes = propTypes;
SharePermalinkButton.defaultProps = defaultProps;

export default React.memo(SharePermalinkButton);
