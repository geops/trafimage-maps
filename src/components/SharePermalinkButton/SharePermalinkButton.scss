.wkp-share-permalink-button {
  position: relative;

  .wkp-tooltip,
  .tm-permalink-field,
  .wkp-arrow-left {
    border: 1px solid lightgray;
  }

  .wkp-tooltip {
    position: fixed;
    background: white;
    width: 400px;
    z-index: 1000;
    height: auto;
    padding: 10px;

    .tm-permalink-field {
      position: relative;
      width: 375px;
      height: 30px;
      padding-left: 5px;

      div[role='button'] {
        border-left: 1px solid lightgray;
        width: 30px;
        display: flex;
        align-items: center;
        justify-content: center;
      }
    }
  }

  .wkp-arrow-left {
    position: absolute;
    left: -6px;
    top: 20px;
    border-width: 0 1px 1px 0;
    display: inline-block;
    padding: 5px;
    transform: rotate(135deg);
    -webkit-transform: rotate(135deg);
    background: white;
  }

  .wkp-close-bt {
    position: absolute;
    right: 0;
    top: 0;
    width: 20px;
    height: 30px;
    display: flex;
    align-items: center;
  }
}
