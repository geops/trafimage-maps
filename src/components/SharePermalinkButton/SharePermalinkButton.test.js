import React from "react";
import { Provider } from "react-redux";

import { render } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import theme from "../../themes/default";
import SharePermalinkButton from ".";

describe("SharePermalinkButton", () => {
  let store;
  test("should match snapshot.", () => {
    store = global.mockStore({
      map: {},
      app: { drawIds: {} },
    });

    const component = render(
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <SharePermalinkButton />
        </Provider>
      </ThemeProvider>,
    );
    expect(component.container.innerHTML).toMatchSnapshot();
  });
});
