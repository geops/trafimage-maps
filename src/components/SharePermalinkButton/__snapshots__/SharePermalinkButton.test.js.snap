// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharePermalinkButton should match snapshot. 1`] = `
<div
  className="wkp-share-permalink-button"
>
  <div
    aria-label="Permalink generieren"
    className="tm-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
    title="Permalink generieren"
  >
    <svg
      focusable={false}
    >
      link_152_large.svg
    </svg>
  </div>
</div>
`;
